### ISSUE TITLE
[API Proposal]: {ReadOnly}Span(ref T) constructor

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#67445]
### Background and motivation

As part of the C# 11 work to enable ref fields, it will be safe to create a {ReadOnly}Span from a ref (with an implicit length of 1), because the compiler will track the lifetime of the ref and ensure the created span can't outlive it.  Creating spans for a single item is quite common, and we should have public constructors for this situation.

This has been proposed by others offline... I agree with it and am just opening the issue for it so we can make forward progress on it.
cc: @jaredpar, @AaronRobinsonMSFT, @tannergooding 

### API Proposal

```diff
namespace System
{
    public ref struct Span<T>
    {
+        public Span(ref T reference);
    }

    public ref struct ReadOnlySpan<T>
    {
+        public ReadOnlySpan(ref T reference);
    }
}
```

### API Usage

```C#
int i = ...;
CallMethodThatTakesASpan(new ReadOnlySpan<T>(ref i));
```

**Comment by stephentoub** [comment]
Examples for all the places in corelib where we create length-1 spans around a single reference are in https://github.com/dotnet/runtime/pull/67447.

**Comment by jaredpar** [comment]
> public ReadOnlySpan(ref T reference);

Believe this should be `in T reference`

**Comment by stephentoub** [comment]
Fixed

**Comment by Sergio0694** [comment]
Is this (the `ReadOnlySpan<T>` constructor) meant to be shipped with an analyzer to disallow rvalue-s as previous discussed for similar cases? Otherwise this would allow eg. `new ReadOnlySpan<int>(5)`, or creating one from a by-value return of some API, with the span actually just pointing to a hidden copy of the value. Unless allowing this is by design, or just not a concern? 🤔

**Comment by stephentoub** [comment]
That's why I'd gone with ref (that, and more generally I really dislike not having to specify that the thing you're passing in is being passed by ref, e.g. `new Span<T>(in value)` instead of `new Span<T>(value)`), but I don't care enough to fight about it.  If you pass in an rvalue, you get what you get I guess.

**Comment by jaredpar** [comment]
> Otherwise this would allow eg. new ReadOnlySpan<int>(5), or creating one from a by-value return of some API, with the span actually just pointing to a hidden copy of the value. Unless allowing this is by design, or just not a concern? 

This is explicitly supported by the language hence it doesn't pose a safety concern. 

> and more generally I really dislike not having to specify that the thing you're passing in is being passed by ref, e.g. I wish you had to write new Span<T>(in value) instead being allowed to write new Span<T>(value)

That is something that we can consider adding to the language. One simple idea that came to my mind as I was reading this is allowing two syntaxes on parameters: 

- `in`: means what it means today
- `ref readonly`: provides a warning at the call site when parameter is not an lvalue

Could also tweak the second idea to be an error but then you're removing your ability to upgrade existing APIs.

**Comment by AaronRobinsonMSFT** [comment]
> ref readonly: provides a warning at the call site when parameter is not an lvalue but otherwise acts like an in parameter

+1

**Comment by tfenise** [comment]
Why is `public Span(in T reference);` also `in`?

**Comment by jaredpar** [comment]
> Why is public Span(in T reference); also in?

That should be `ref` otherwise it violates `readonly` rules

**Comment by stephentoub** [comment]
> Why is public Span(in T reference); also in?

Editing in haste.

**Comment by tannergooding** [comment]
> That is something that we can consider adding to the language. One simple idea that came to my mind as I was reading this is allowing two syntaxes on parameters:

Yes please. I would love this and it would solve one of the biggest issues I have with `in` (the other is changing a signature from `ref` to `in` is a source breaking change; if we allowed that for `ref` to `ref readonly` then its even better as we could update already shipped APIs).

**Comment by terrajobst** [comment]
* Looks good as proposed
* Once C# supports `ref readonly` (csharplang#6010) we should change the constructor for `ReadOnlySpan<T>` to use that instead.

```C#
namespace System;

public partial ref struct Span<T>
{
    public Span(ref T reference);
}
public partial ref struct ReadOnlySpan<T>
{
    public ReadOnlySpan(in T reference);
}
```

**Comment by stephentoub** [comment]
(Just noting we should not expose these until the C# safety work is completed.)

**Comment by halter73** [comment]
Are there really safety concerns with passing an rvalue to the ReadOnlySpan constructor or is just inefficient? Does it basically just box the value?

**Comment by tannergooding** [comment]
Its not, IMO, a safety concern, but rather one of intent. Basically there are two considerations around `readonly byrefs`. 

In one hand you have cases where it being a `readonly byref` is really an "implementation detail", it largely only exists as a perf optimization. Most of the places we use it in the BCL (or that you see `const T&` in C/C++) are similar.

On the other hand, you have cases where it being a `readonly byref` is a very important detail, generally because the API is directly dealing with the concept of memory. Consider for example `Unsafe.IsNullRef(ref T)`. While there is nothing "conceptually" wrong with passing an `rvalue` here and `Unsafe.IsNullRef(5)` will do "the right thing", it's likely an error/mistake on the end of the consumer.

In the same way, `Span` and `ReadOnlySpan` are intentionally capturing and tracking the passed in `byref`. There is nothing conceptually wrong with `new ReadOnlySpan(5)`, but it represents a potential error/problem case due to the API dealing explicitly with the concept of memory. The cases where a user might want an `rvalue` are cases like `params ReadOnlySpan<T>` and in that case  other syntax should handle the creation of said span.

**Comment by stephentoub** [comment]
The safety work I was referring to above wasn't about rvalues; it was about the C# compiler tracking refs.  If with the C# compiler today we were to expose these constructors, you could do:
```C#
public Span<int> DoNotDoThis()
{
    int i = 0;
    return new Span<int>(ref i);
}
```
the compiler would happily compile that, and it would be really bad.

**Comment by terrajobst** [comment]
> the compiler would happily compile that, and it would be really bad.

I don't always peek into the stack, but when I do, I make sure it's into a stack frame that was unwound

**Comment by deeprobin** [comment]
Does this issue include changes in the JIT or VM or just exposing the private/internal constructors + appropriate tests?

**Comment by tannergooding** [comment]
This can't be implemented until the relevant language changes are in.

**Comment by jaredpar** [comment]
@deeprobin this issue just covers the API changes around `Span / ReadOnlySpan`. There are many other changes that need to happen to support `ref` fields. The overarching issue tracking the work is https://github.com/dotnet/runtime/issues/63768

**Comment by deeprobin** [comment]
Thank you :+1:
